File: Artificial Intelligence/Homework/Move.py
#!/usr/bin/python
from Coordinates import Coordinates
from brick import Brick

UP = 0
DOWN = 1
LEFT = 2
RIGHT = 3

class Move(object):

    # ========================================
    # @ Description - Movement object representation
    # ========================================
    def __init__(self, piece=-1, direction=-1, coords=None):
        self.piece = piece
        self.direction = direction
        self.coords = coords

    # ========================================
    # @ Description - applies this move to state
    # @ param[in] <state> - the state in question
    # @ return <void>
    # ========================================
    def apply_move(self, state):
        coord_list = self.get_piece_coords(state, self.piece)
        brick = Brick()
        brick.set_height_width(coord_list)

        valid = brick.check_move_valid(state, self)

        if valid == True:
            if self.direction == UP:
                for coord in coord_list:
                    x = coord.x
                    y = coord.y
                    if state.board[x - 1][y] == 0:
                        state.swap_board_elems(x, y, x - 1, y)
                    elif state.board[x - 1][y] == -1:
                        state.replace_zero(x, y, x - 1, y)

            if self.direction == DOWN:
                for coord in coord_list:
                    x = coord.x
                    y = coord.y
                    if state.board[x + 1][y] == 0:
                        state.swap_board_elems(x, y, x + 1, y)
                    elif state.board[x + 1][y] == -1:
                        state.replace_zero(x, y, x + 1, y)

            if self.direction == LEFT:
                for coord in coord_list:
                    x = coord.x
                    y = coord.y
                    if state.board[x][y - 1] == 0:
                        state.swap_board_elems(x, y, x, y - 1)
                    elif state.board[x][y - 1] == -1:
                        state.replace_zero(x, y, x, y - 1)

            # Have to reverse direction for right, as coordinates 
            # are naturally ordered from left to right. 
            # swapping right does not follow this behavior
            if self.direction == RIGHT:
                for coord in reversed(coord_list):
                    x = coord.x
                    y = coord.y
                    if state.board[x][y + 1] == 0:
                        state.swap_board_elems(x, y, x, y + 1)
                    elif state.board[x][y + 1] == -1:
                        state.replace_zero(x, y, x, y + 1)

    def check_move_valid(self, state):
        coord_list = self.get_piece_coords(state, self.piece)

        valid = True
        for coord in coord_list:
            x = coord.x
            y = coord.y

            if self.direction == UP:
                if state.board[x - 1][y] != 0:
                    valid = False
            if self.direction == DOWN:
                if state.board[x + 1][y] != 0:
                    valid = False
            if self.direction == LEFT:
                if state.board[x][y - 1] != 0\
                and state.board[x][y - 1] != coord.value:
                    valid = False
            if self.direction == RIGHT:
                if state.board[x][y + 1] != 0:
                    valid = False

    # ========================================
    # @ Description - applies move and clones state
    # @ param[in] <state> - the state in question
    # @ return <GameState>
    # ========================================
    def apply_move_cloning(self, state):
        new_state = state.copy()
        self.apply_move(state)
        return new_state

    # ========================================
    # @ Description - retrieves piece coords
    # @ param[in] <state> - the state to search
    # @ param[in] <piece> - the piece to look for
    # @ return <list<Coordinates>>
    # ========================================
    def get_piece_brick(self, state, piece):
        coord_list = []

        for i in range(0, state.rows):
            for j in range(0, state.cols):
                if state.board[i][j] == piece:
                    coord_list.append(Coordinates(piece, i, j))

        brick = Brick()
        brick.set_height_width(coord_list)

        return brick

    def get_piece_coords(self, state, piece):
        coord_list = []

        for i in range(0, state.rows):
            for j in range(0, state.cols):
                if state.board[i][j] == piece:
                    coord_list.append(Coordinates(piece, i, j))

        return coord_list

    # ========================================
    # @ Description - gets all zero coordinates
    # @ param[in] <state> - the state to search
    # @ return <list<Coordinates>>
    # ========================================
    def get_zero_coords(self, state):
        zero_list = []

        for i in range(0, state.rows):
            for j in range(0, state.cols):
                if state.board[i][j] == 0:
                    zero_list.append(Coordinates(0, i, j))

        return zero_list

    # ========================================
    # @ Description - gets all moves for piece
    # @ param[in] <state> - the state to search
    # @ return <list<Move>>
    # ========================================
    def get_piece_moves(self, state, piece):
        piece_coords = self.get_piece_coords(state, piece)

        move_list = []

        for p_coords in piece_coords:
            if p_coords.up_free(state):
                move_list.append(Move(p_coords.value, UP, p_coords))
            elif p_coords.down_free(state):
                move_list.append(Move(p_coords.value, DOWN, p_coords))
            elif p_coords.left_free(state):
                move_list.append(Move(p_coords.value, LEFT, p_coords))
            elif p_coords.right_free(state):
                move_list.append(Move(p_coords.value, RIGHT, p_coords))

        brick = self.get_piece_brick(state, piece)

        for move in move_list:
            if brick.check_move_valid(state, move) == False:
                move_list.remove(move)

        return move_list

    # ========================================
    # @ Description - gets all available moves
    # @ param[in] <state> - the state in question
    # @ return <list<Move>>
    # ========================================
    def get_all_moves(self, state):
        pieces = state.get_all_pieces()
        all_move_list = []

        for piece in pieces:
            piece_move_list = self.get_piece_moves(state, piece)

            if piece_move_list is None:
                all_move_list.append(str(piece) + " has no moves")
            else:
                for move in piece_move_list:
                    all_move_list.append(move)

        return all_move_list

    # ========================================
    # @ Description - prints out a move
    # ========================================
    def string(self):
        if self.direction == UP:
            print("(" + str(self.piece) + ", up)")
        elif self.direction == DOWN:
            print("(" + str(self.piece) + ", down)")
        elif self.direction == LEFT:
            print("(" + str(self.piece) + ", left)")
        elif self.direction == RIGHT:
            print("(" + str(self.piece) + ", right)")

File: Artificial Intelligence/Side Projects/Move.cpp
#include "Move.h"

class Cooridnates;
class GameState;

#include <iostream>
#include <map>

//============================================================
/// @ Description - generates all moves for a given piece
///
/// @ param[in] <state> - the board state
/// @ param[in] <piece> - the piece to consider
/// @ return<deque<move>> - the move_list
//============================================================
deque<Move> Move::get_piece_moves(const GameState &state, int piece) const
{
	deque<Move> move_list;
	Coordinates piece_coords;

	deque<Coordinates> zero_list = Coordinates::zero_list(state, piece);
	cout << "zero list size " << zero_list.size() << endl;
	for(int i = 0; i < state.get_rows(); i++)
	{
		for(int j = 0; j < state.get_cols(); j++)
		{
			cout << "state " << &state << endl;
			int elem = state.get_board_elem(i, j);
			cout << "piece move error2\n";
			if(state.get_board_elem(i, j) == piece)
			{
				piece_coords.set_value(piece);
				piece_coords.set_x(i);
				piece_coords.set_y(j);

				zero_list.push_back(piece_coords);
			}
		}
	}

	for(int i = 0; i < zero_list.size(); i++)
	{
		Move move;
		if (piece_coords.is_up(zero_list.at(i)))
			move = Move(piece, Move::UP, piece_coords);
		else if (piece_coords.is_down(zero_list.at(i)))
			move = Move(piece, Move::DOWN, piece_coords);
		else if (piece_coords.is_left(zero_list.at(i)))
			move = Move(piece, Move::LEFT, piece_coords);
		else if (piece_coords.is_right(zero_list.at(i)))
			move = Move(piece, Move::RIGHT, piece_coords);

		move_list.push_back(move);
	}

	return move_list;
}

//============================================================
/// @ Description - returns all moves for a given board state
///
/// @ param[in] <state> - the state to consider
/// @ return <deque<Move>> - a deque of all moves
//============================================================
deque<Move> Move::get_all_moves(const GameState &state) const
{
	cout << "ge here\n";
	deque<int> pieces = state.get_all_pieces();

	deque<Move> piece_move_list;

	for(int i = 0; i < pieces.size(); i++)
	{
		cout << "get here\n";
		deque<Move> piece_move_list = Move::get_piece_moves(state, pieces.at(i));

		// cout << "Adj List size " << zero_list.size() << endl;
		// for(map<int, Coordinates>::iterator iter = adjacent_list.begin(); iter != adjacent_list.end(); ++iter)
		// {
		// 	Move new_move = Move(pieces.at(i), iter->first, iter->second);
		// 	this->move_list.push_back(new_move);
		// }
	}

	return piece_move_list;
}

//============================================================
/// @ Description - apply_move applies the move to the state
///
/// @ param[in] <state> - the state to apply move to
/// @ param[in] <move> - the move to apply_move
/// @ return <void>
//============================================================
void Move::apply_move(const GameState &state, Move move) const
{
	int target = move.get_piece();
	deque<Coordinates> coordinate_list;

	for (int i = 0; i < state.get_rows(); i++)
		for(int j = 0; j < state.get_cols(); j++)
			if(state.get_board_elem(i, j) == target)
				coordinate_list.push_back(Coordinates(target, i, j));

	// shift the piece the appropriate amount
	for(int i = 0; i < coordinate_list.size(); i++)
	{
		Coordinates coord = coordinate_list.at(i);

		int x = coord.get_x();
		int y = coord.get_y();

		if(state.get_board_elem(x, y) == move.get_piece()
		|| state.get_board_elem(x, y) == GameState::EMPTY)
		{
			if(move.get_direction() == Move::UP
			&& state.get_board_elem((x - 1), y) == GameState::EMPTY)
				state.swap_board_elems(x, y, x - 1, y);

			if(move.get_direction() == Move::DOWN
			&& state.get_board_elem((x + 1), y) == GameState::EMPTY)
				state.swap_board_elems(x, y, x + 1, y);

			if(move.get_direction() == Move::LEFT
			&& state.get_board_elem(x, (y - 1)) == GameState::EMPTY)
				state.swap_board_elems(x, y, x, y - 1);

			if(move.get_direction() == Move::RIGHT
			&& state.get_board_elem(x, (y + 1)) == GameState::EMPTY)
				state.swap_board_elems(x, y, x, y + 1);
		}
	}
}

//============================================================
/// @ Description - clones the given state and applies the move
///
/// @ param[in] <state> - the state to clone and apply move to
/// @ param[in] <move> - the move to apply
/// @ return <GameState> - a new state with the applied move
//============================================================
GameState Move::apply_move_cloning(const GameState &state, Move move) const
{
	GameState new_state = GameState(state);

	move.apply_move(new_state, move);

	return new_state;
}

//============================================================
/// @ Description - print move formats and prints the move
///
/// @ return <void>
//============================================================
void Move::print_move()
{
	if(this->direction == Move::UP)
		cout << "(" << this->piece << ", " << "up" << ")" << endl;
	else if(this->direction == Move::DOWN)
		cout << "(" << this->piece << ", " << "down" << ")" << endl;
	else if(this->direction == Move::LEFT)
		cout << "(" << this->piece << ", " << "left" << ")" << endl;
	else if(this->direction == Move::RIGHT)
		cout << "(" << this->piece << ", " << "right" << ")" << endl;
}

//============================================================
/// @ Description - prints all moves for state
///
/// @ param[in] <state> - the state in question
/// @ return <void>
//============================================================
void Move::print_all_moves(const GameState state) const
{
	Move move;
	cout << "Error here\n";
	deque<Move> moves = move.get_all_moves(state);

	for(int i = 0; i < moves.size(); i++)
		moves.at(i).print_move();
}
